File: Model Branch/C_cartesiano.cpp
//
// Created by luca on 19/12/17.
//

#include "C_cartesiano.h"
#include "C_polare.h"
#include <cmath>

C_cartesiano::C_cartesiano(double r, double i):reale(r),immaginaria(i){};

C_cartesiano::C_cartesiano(const C_cartesiano& c):reale(c.reale),immaginaria(c.immaginaria){};

C_cartesiano::C_cartesiano(std::string s){
  auto pos=s.find('i');
  std::string::size_type size=0;
  reale=0;
  immaginaria=0;
  if(pos==-1)
    reale=std::stod(s);
  else if(pos==s.length()-1)
    immaginaria=(pos==0?1:std::stod(s.substr(size)));
  else
      throw syntax_exception("Invalid value"); //gestire eccezione syntax error
}

C_cartesiano::C_cartesiano():reale(0),immaginaria(0){};

C_cartesiano* C_cartesiano::operator+ (const Numero* n)const {
    if(dynamic_cast<const C_cartesiano*>(n)){
      const C_cartesiano* c=static_cast<const C_cartesiano*>(n);
      return new C_cartesiano(reale+c->reale,immaginaria+c->immaginaria);
    }
    else if(dynamic_cast<const C_polare*>(n)){
      const C_polare* cp=static_cast<const C_polare*>(n);
      const C_cartesiano* aux=static_cast<const C_cartesiano*>(cp->converti());
      delete cp;
      return this->operator+(aux);
    }
    else
    throw logic_exception("Tipo incompatibile");    //gestire eccezione di tipo incompatibile
}

C_cartesiano* C_cartesiano::operator- (const Numero* n)const {
  if(dynamic_cast<const C_cartesiano*>(n)){
    const C_cartesiano* c=static_cast<const C_cartesiano*>(n);
    return new C_cartesiano(reale-c->reale,immaginaria-c->immaginaria);
  }
  else if(dynamic_cast<const C_polare*>(n)){
    const C_polare* cp=static_cast<const C_polare*>(n);
    const C_cartesiano* aux=static_cast<const C_cartesiano*>(cp->converti());
    delete cp;
    return this->operator-(aux);
  }
  else
      throw logic_exception("Tipo incompatibile");    //gestire eccezione di tipo incompatibile
}

C_cartesiano* C_cartesiano::operator* (const Numero* n)const {
  if(dynamic_cast<const C_cartesiano*>(n)){
    const C_cartesiano* c=static_cast<const C_cartesiano*>(n);
    return new C_cartesiano(reale*c->reale-immaginaria*c->immaginaria,immaginaria*c->reale+reale*c->immaginaria);
  }
  else if(dynamic_cast<const C_polare*>(n)){
    const C_polare* cp=static_cast<const C_polare*>(n);
    const C_cartesiano* aux=static_cast<const C_cartesiano*>(cp->converti());
    delete cp;
    return this->operator*(aux);
  }
  else
      throw logic_exception("Tipo incompatibile");    //gestire eccezione di tipo incompatibile
}

C_cartesiano* C_cartesiano::operator/ (const Numero* n)const {
  if(dynamic_cast<const C_cartesiano*>(n)){
    const C_cartesiano* c= static_cast<const C_cartesiano*>(n);
    return new C_cartesiano((reale*c->reale+immaginaria*c->immaginaria)/(pow(reale,2)+pow(c->immaginaria,2)),(immaginaria*c->reale-reale*c->immaginaria)/(pow(reale,2)+pow(c->immaginaria,2)));
  }
  else if(dynamic_cast<const C_polare*>(n)){
    const C_polare* cp= static_cast<const C_polare*>(n);
    const C_cartesiano* aux= static_cast<const C_cartesiano*>(cp->converti());
    delete cp;
    return this->operator+(aux);
  }
  else
      throw logic_exception("Tipo incompatibile");    //gestire eccezione di tipo incompatibile
}

//C_cartesiano& C_cartesiano::operator=(const Dato& d){
//    try{
//        auto aux= dynamic_cast<const C_cartesiano&>(d);
//        reale=aux.reale;
//        immaginaria=aux.immaginaria;
//        return *this;
//    }
//    catch (const std::bad_cast &error){
//        std::cout << "tipi incompatibili" << std::endl;
//    }
//}

bool C_cartesiano::operator== (const Dato& d)const{
    try {
        auto aux = dynamic_cast<const C_cartesiano &>(d);
        return this->reale == aux.reale && this->immaginaria == aux.immaginaria;
    }
    catch(const std::bad_cast& error){
        return false;
    }
}

double C_cartesiano::getReale() const {return reale;}
double C_cartesiano::getImmaginaria() const {return immaginaria;}

C_cartesiano* C_cartesiano::coniugato() const{
    return new C_cartesiano(reale, immaginaria*-1);
}

Complesso* C_cartesiano::converti() const{
    double f=sqrt(pow(reale,2)+pow(immaginaria,2));
    double m=rad_to_deg(atan(immaginaria/reale));
    return new C_polare(f,m);
}

void C_cartesiano::stampa(std::ostream& os)const {
    std::cout<<reale;
    if(immaginaria<0)std::cout<< immaginaria;
    else std::cout<<"+"<<immaginaria;
    std::cout<<"i";
}

std::ostream& operator<<(std::ostream& os, const C_cartesiano& n){
    n.stampa(os);
    return os;
}

File: Java/CCartesiano.java
//classe concreta CCartesiano, derivata da Complesso
//rappresenta un numero complesso in forma cartesiana

//OBJECT -> equals()
//OBJECT -> toString()

//NUMERO -> somma()   ok
//NUMERO -> differenza()    ok
//NUMERO -> moltiplicazione()   ok
//NUMERO -> divisione()   ok

//COMPLESSO -> converti()   ok
//COMPLESSO -> coniugato()    ok

//CCARTESIANO ->costruttore di default    ok
//CCARTESIANO ->costruttore da due double   ok
//CCARTESIANO ->getReale()    ok
//CCARTESIANO ->getImmaginaria()    ok


class CCartesiano extends Complesso{
  private double reale;
  private double immaginaria;

  public double getReale(){
    return reale;
  }

  public double getImmaginaria(){
    return immaginaria;
  }

  public CCartesiano(double r, double i){
    reale = r;
    immaginaria = i;
  }

  public CCartesiano(){
    this(0, 0);
  }

  public Complesso converti(){
    double f = Math.sqrt(Math.pow(reale, 2) + Math.pow(immaginaria, 2));
    double m = Math.toDegrees(Math.atan(immaginaria / reale));
    CPolare result = new CPolare(f, m);
    return result;
  }

  public Complesso coniugato(){
    CCartesiano result = new CCartesiano(reale, -immaginaria);
    return result;
  }

  public Numero somma(Numero n) throws KalkException{
    if(n instanceof CCartesiano){
      CCartesiano aux = (CCartesiano) n;
      CCartesiano result = new CCartesiano(reale + aux.reale, immaginaria + aux.immaginaria);
      return result;
    }
    else if(n instanceof CPolare){
      CPolare cp = (CPolare) n;
      CCartesiano aux = (CCartesiano) (cp.converti());

      Numero result = somma(aux);
      return result;
    }
    else{
      throw new KalkException("tipi incompatibili", false);
      //return this;
    }
  }

  public Numero differenza(Numero n) throws KalkException{
    if(n instanceof CCartesiano){
      CCartesiano aux = (CCartesiano) n;
      CCartesiano result = new CCartesiano(reale - aux.reale, immaginaria - aux.immaginaria);
      return result;
    }
    else if(n instanceof CPolare){
      CPolare cp = (CPolare) n;
      CCartesiano aux = (CCartesiano) (cp.converti());

      Numero result = differenza(aux);
      return result;
    }
    else{
      throw new KalkException("tipi incompatibili", false);
    }
  }

  public Numero moltiplicazione(Numero n) throws KalkException{
    if(n instanceof CCartesiano){
      CCartesiano aux = (CCartesiano) n;
      CCartesiano result = new CCartesiano((reale * aux.reale) - (immaginaria * aux.immaginaria), (immaginaria * aux.reale) + (reale * aux.immaginaria));
      return result;
    }
    else if(n instanceof CPolare){
      CPolare cp = (CPolare) n;
      CCartesiano aux = (CCartesiano) (cp.converti());

      Numero result = moltiplicazione(aux);
      return result;
    }
    else{
      throw new KalkException("tipi incompatibili", false);
    }
  }

  public Numero divisione(Numero n) throws KalkException{
    if(n instanceof CCartesiano){
      CCartesiano aux = (CCartesiano) n;
      double r = ((reale * aux.reale) + (immaginaria * aux.immaginaria)) / (Math.pow(reale, 2) + Math.pow(aux.immaginaria, 2));
      double i = ((immaginaria * aux.reale) - (reale * aux.immaginaria)) / (Math.pow(reale, 2) + Math.pow(aux.immaginaria, 2));
      CCartesiano result = new CCartesiano(r, i);
      return result;
    }
    else if(n instanceof CPolare){
      CPolare cp = (CPolare) n;
      CCartesiano aux = (CCartesiano) (cp.converti());

      Numero result = divisione(aux);
      return result;
    }
    else{
    throw new KalkException("tipi incompatibili", false);
    }
  }

  public boolean equals(Object obj) throws KalkException{
    if(!(obj instanceof CCartesiano)){
      throw new KalkException("cast fallito", false);
    }

    CCartesiano aux = (CCartesiano) obj;

    if((reale == aux.reale) && (immaginaria == aux.immaginaria)){
      return true;
    }
    else{
      return false;
    }
  }

  public String toString(){
    String result = new String(reale + " " + ((immaginaria < 0) ? "-" : "+") + " i" + Math.abs(immaginaria) + "\r\n");
    return result;
  }
}
